1. Реализована првоерка статуса при переходе.

mramazanov opened the pull request:
2. Реализована отправка нотификации при создании или изменении задачи
3. Реализована проверка роли пользователя при создании или изменения задачи
4. Доработаны тесты

mramazanov committed:
1. Реализована првоерка статуса при переходе.
2. Реализована отправка нотификации при создании или изменении задачи
3. Реализована проверка роли пользователя при создании или изменения задачи
4. Доработаны тесты

KiryakovDenis commented on src/main/java/ru/javajabka/taskservice/service/TaskService.java:
```diff
@@ -33,7 +38,15 @@ public Task create(final TaskRequestDTO taskRequest) {
                 .build();
 
-        return taskServiceRepository.create(task);
+        Task createdTask = taskServiceRepository.create(task);
```
Мне кажется можно повысить читабельность кода, если вынести в отдельный метод формирование экземпляра EventDTO и вызов notificationProducer.send(eventDTO);

Nokichi replied:
++

mramazanov replied:
Тут оставил именно создание обекта.
Там ниже в коде сделал для if блоков отдельные 2 метода

mramazanov replied:
Поправил

KiryakovDenis commented on src/main/java/ru/javajabka/taskservice/service/TaskService.java:
```diff
@@ -67,6 +85,85 @@ public List<Task> getAll(
+        LocalDateTime changeTime = LocalDateTime.now();
+
+        if (!oldTask.getTitle().equals(newTask.getTitle())) {
+            EventDTO event = EventDTO.builder()
```
Много повторяющегося кода. Можно упаковать создание и отправку события в метод с  сигнатурой {[Имя события],[ИД задачи],[Старое значение], [Новое значение]}. Если в каждом if будет вызов одного метода будет выглядеть проще

Nokichi replied:
++

mramazanov replied:
Поправил

KiryakovDenis commented on src/main/java/ru/javajabka/taskservice/service/TaskService.java:
```diff
@@ -67,6 +85,85 @@ public List<Task> getAll(
+    private void checkStatus(TaskStatus fromStatus, TaskStatus toStatus) {
+        boolean deniedStatus = false;
+
+        if (fromStatus == TaskStatus.TO_DO && toStatus == TaskStatus.DONE) {
```
Для сопровождения, лютый вариант реализации, если граф переходов состояний будет сложнее, то количество if рискует сильно увеличиться. Я бы рассмотрел какую-нибудь структуру данных, и анализировал бы ее. Как вариант можно взять Map<Status, Set<Status>>.

mramazanov replied:
Да согласен. При увеличении или большом количестве условий можно применить паттерн Rule или стратегию.
Не стал тут это делать, так как кейс пока очень простой.

Nokichi commented on src/test/java/ru/javajabka/taskservice/TaskServiceTest.java:
```diff
@@ -198,4 +202,4 @@ private TaskUpdateDTO buildTaskUpdateDTO(Long id, String title, String descripti
                 .build();
     }
-}
+}
```
а тесты на проверку переходов состояний не добавлял? Может не заметил...

mramazanov committed:
Фиксы замечаний

mramazanov committed:
Фиксы замечаний v2

mramazanov committed:
Фиксы замечаний v3

gmnpr commented on src/main/java/ru/javajabka/taskservice/service/TaskService.java:
```diff
@@ -4,21 +4,26 @@
-import java.util.List;
-import java.util.Optional;
+import java.time.LocalDateTime;
+import java.util.*;
```
не используем запись импортов с *

mramazanov replied:
Сделал настройку в intel idea count до 100, думаю не наберётся столько импортов)

gmnpr commented on src/main/java/ru/javajabka/taskservice/service/TaskService.java:
```diff
@@ -67,6 +79,42 @@ public List<Task> getAll(
         return taskServiceRepository.getAll(status, assignee);
     }
 
+    private void sendEvent(final Task oldTask, final Task newTask) {
```
немножко оверинженеринг конечно, ну да ладно

gmnpr commented on src/main/java/ru/javajabka/taskservice/service/TaskService.java:
```diff
@@ -42,11 +49,15 @@ public Task getById(final Long id) {
         Task foundTask = taskServiceRepository.getById(taskUpdateDTO.getId());
 
+        if (taskUpdateDTO.getStatus() != null) {
+            checkStatus(foundTask.getStatus(), taskUpdateDTO.getStatus());
```
почему это не внутри валидации?)

gmnpr commented on src/main/java/ru/javajabka/taskservice/service/TaskService.java:
```diff
@@ -91,18 +139,41 @@ private void validate(final TaskRequestDTO taskRequest) {
             throw new BadRequestException("Введите идентификатор ответственного больше нуля");
         }
-        userService.checkUserId(List.of(taskRequest.getAuthor(), taskRequest.getAssignee()));
+        List<User> users = userService.checkUserId(List.of(taskRequest.getAuthor(), taskRequest.getAssignee()));
```
может это засунуть внутрь метода checkUserRole?

gmnpr commented on src/main/java/ru/javajabka/taskservice/service/TaskService.java:
```diff
@@ -91,18 +139,41 @@ private void validate(final TaskRequestDTO taskRequest) {
             throw new BadRequestException("Введите идентификатор ответственного больше нуля");
         }
-        userService.checkUserId(List.of(taskRequest.getAuthor(), taskRequest.getAssignee()));
+        List<User> users = userService.checkUserId(List.of(taskRequest.getAuthor(), taskRequest.getAssignee()));
```
это странно, что метод checkUserId возвращается список пользователей

mramazanov replied:
Успел поменять много чего, но только не метод))

gmnpr commented on src/main/java/ru/javajabka/taskservice/service/TaskService.java:
```diff
@@ -91,18 +139,41 @@ private void validate(final TaskRequestDTO taskRequest) {
     }
+
+    private void checkUserRole(List<User> users, Long authorId) {
+        users.stream().filter(user -> user.getId().equals(authorId)).findFirst().ifPresent(user -> {
```
если я не менеджер я смогу поменять описание задачи?

mramazanov replied:
Тут наверное в заблуждение вводит чуть название переменной authorId(нужно editorId)
Да, проверил, если указать id пользователя с ролью отличной от менеджера то получим ошибку.

gmnpr replied:
проверь требования
там 
>Только MANAGER может назначать менять исполнителя у задачи

mramazanov replied:
Согласно задаче сделал изначально, запутался в твоем вопросе)

mramazanov committed:
1. Сделал batсh отправку событий разом
2. Исправил баг в отправке событий, при получении ошибки сохранения в БД - событие теперь не отправляется
3. Испарвления после Code Review

gmnpr commented on src/main/java/ru/javajabka/taskservice/service/TaskService.java:
```diff
@@ -91,18 +180,40 @@ private void validate(final TaskRequestDTO taskRequest) {
-        if (taskUpdateDTO != null && taskUpdateDTO.getAssignee() != null) {
-            userService.checkUserId(List.of(taskUpdateDTO.getAssignee()));
+        if (taskUpdateDTO.getAssignee() != null) {
+            List<User> user = userService.checkAndGetUsers(List.of(taskUpdateDTO.getAssignee(), authorId));
```
user или users?)

mramazanov committed:
Правки мелкие полсе Code Review
